### ISSUE TITLE
Proposal: Make the project generalized

### ISSUE THREAD

**Issue by Naddiseo** [benjamn/ast-types#57]
Currently, this project is geared towards ES, but since the interface is so nice for building asts, I'd like to use it in other parsing projects.
Would it be possible to make the project more generalized so that it doesn't load all the types in `def/*` unless requested?

**Comment by benjamn** [comment]
I still think this is a great idea, by the way. Hope to work on it for the next minor version (0.8.0).

**Comment by jamestalmage** [comment]
bump!
`0.9.0` maybe?

Personally, I think the core components should be split out into another library (`ast-types-base` for sake of argument).

I like the syntax, but I think you should pass actual modules (or a custom function) instead of strings. The problem with strings would be accurately resolving them (especially if you do split into a separate module):

```js
var types = require('ast-types-base')(
  require('ast-types/def/core'),
  require('./myCustomDefs'),
  function(types) {
   types.def('MyType')
     //...
  }
);
```

I do like the `fork` option discussed, and that seems appropriate for `ast-types`, but not `ast-types-base` as I'm proposing. Perhaps strings would make sense there for specifying some of the built in stuff.

This would be the equivalent, less verbose version of above:

```js
require('ast-types').fork(
  'core',
  require('./myCustomDefs'),
  function(types) {
   types.def('MyType')
     //...
  }
);
```

One problem with the `fork` method as currently being discussed would be with `browserify` builds. Browserify will pick up and package everything in `def/*`, even if you aren't using any of it and creating an entirely custom AST.

**Comment by jamestalmage** [comment]
@benjamn - anything I can do to move this forward?

**Comment by bnjmnt4n** [comment]
Should probably be closed since #145 was merged nearly 2.5 years (!!!) ago.

**Comment by benjamn** [comment]
Agreed, though I should mention the move to TypeScript (#300) has complicated this story, since the generated TypeScript declarations are fundamentally less modular/dynamic than the forking system.
